UniversalAlgebra/Homomorphisms/Setoid.lagda

```diff
@@ -0,0 +1,307 @@
+ is-homomorphism h = ∀ 𝑓  →  compatible-op-map 𝑓 h
+
+ -- The type of homomorphisms from `𝑨` to `𝑩`.
+ hom : Type(𝓞 ⊔ 𝓥 ⊔ α ⊔ β)
+ hom = Σ (A → B) is-homomorphism
+
+open PE.≡-Reasoning
+open PE renaming (cong to ≡-cong)
+
+module _ {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)  -- (explicit 𝑨)
+         {β ρᵇ : Level} {𝑩 : SetoidAlgebra β ρᵇ}  -- (implicit 𝑩)
+         {γ ρᶜ : Level} (𝑪 : SetoidAlgebra γ ρᶜ)  -- (explicit 𝑪)
+         where
+ private
+  A = 𝕌[ 𝑨 ]  -- carrier of domain of 𝑨
+  B = 𝕌[ 𝑩 ]
+  C = 𝕌[ 𝑪 ]
+
+ -- The composition of homomorphisms is again a homomorphism.
+ ∘-hom : hom 𝑨 𝑩  →  hom 𝑩 𝑪  →  hom 𝑨 𝑪
+ ∘-hom (g , ghom) (h , hhom) = h ∘ g , Goal where
+
+  Goal : ∀ 𝑓 a → (h ∘ g)((𝑓 ̂ 𝑨) a) ≡ (𝑓 ̂ 𝑪)(h ∘ g ∘ a)
+  Goal 𝑓 a = (h ∘ g)((𝑓 ̂ 𝑨) a)     ≡⟨ ≡-cong h ( ghom 𝑓 a ) ⟩
+             h ((𝑓 ̂ 𝑩)(g ∘ a))     ≡⟨ hhom 𝑓 ( g ∘ a ) ⟩
+             (𝑓 ̂ 𝑪)(h ∘ g ∘ a)     ∎
+
+
+ ∘-is-hom : {f : A → B}{g : B → C}
+  →         is-homomorphism 𝑨 𝑩 f → is-homomorphism 𝑩 𝑪 g
+            -------------------------------------------------
+  →         is-homomorphism 𝑨 𝑪 (g ∘ f)
+
+ ∘-is-hom {f} {g} fhom ghom = ∥ ∘-hom (f , fhom) (g , ghom) ∥
+
+module _ {α ρ : Level} where
+
+ -- the identity homs
+ 𝒾𝒹 : (𝑨 : SetoidAlgebra α ρ) → hom 𝑨 𝑨
+ 𝒾𝒹 _ = id , λ 𝑓 a → refl
+
+ open Level
+
+ -- the lift hom
+ 𝓁𝒾𝒻𝓉 : {β : Level}(𝑨 : SetoidAlgebra α ρ) → hom 𝑨 (Lift-SetoidAlg 𝑨 β)
+ 𝓁𝒾𝒻𝓉 _ = lift , (λ 𝑓 a → refl)
+
+ -- the lower hom
+ 𝓁ℴ𝓌ℯ𝓇 : {β : Level}(𝑨 : SetoidAlgebra α ρ) → hom (Lift-SetoidAlg 𝑨 β) 𝑨
+ 𝓁ℴ𝓌ℯ𝓇 _ = (lower , λ 𝑓 a → refl)
+
+module _ {α ρᵃ : Level} {𝑨 : SetoidAlgebra α ρᵃ}
+         (ℓᵃ : Level)
+         {β ρᵇ : Level}{𝑩 : SetoidAlgebra β ρᵇ}
+         (ℓᵇ : Level)
+         where
+ open Level
+
+ Lift-hom : hom 𝑨 𝑩  →  hom (Lift-SetoidAlg 𝑨 ℓᵃ) (Lift-SetoidAlg 𝑩 ℓᵇ)
+
+ Lift-hom (f , fhom) = lift ∘ f ∘ lower , Goal
+  where
+  lA lB : SetoidAlgebra _ _
+  lA = Lift-SetoidAlg 𝑨 ℓᵃ
+  lB = Lift-SetoidAlg 𝑩 ℓᵇ
+
+  lABh : is-homomorphism lA 𝑩 (f ∘ lower)
+  lABh = ∘-is-hom lA {𝑩 = 𝑨}  𝑩 {lower}{f} (λ _ _ → refl) fhom
+
+  Goal : is-homomorphism lA lB (lift ∘ (f ∘ lower))
+  Goal = ∘-is-hom lA {𝑩 = 𝑩} lB {f ∘ lower}{lift} lABh λ _ _ → refl
+
+
+-- Monomorphisms and epimorphisms
+module _ {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)
+         {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ)
+         where
+
+ private
+  A = 𝕌[ 𝑨 ]  -- carrier of domain of 𝑨
+  B = 𝕌[ 𝑩 ]
+
+ is-monomorphism : (A → B) → Type _
+ is-monomorphism g = is-homomorphism 𝑨 𝑩 g × IsInjective g
+
+ mon : Type(𝓞 ⊔ 𝓥 ⊔ α ⊔ β)
+ mon = Σ[ g ∈ (A → B) ] is-monomorphism g
+
+ is-epimorphism : (A → B) → Type _
+ is-epimorphism g = is-homomorphism 𝑨 𝑩 g × IsSurjective g
+
+ epi : Type _
+ epi = Σ[ g ∈ (A → B) ] is-epimorphism g
+
+
+-- The "hom reduct" of a mon
+-- N.B. 𝑨 explicit, 𝑩 implicit
+module _ {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)
+         {β ρᵇ : Level} {𝑩 : SetoidAlgebra β ρᵇ} where
+
+ mon-to-hom : mon 𝑨 𝑩 → hom 𝑨 𝑩
+ mon-to-hom ϕ = ∣ ϕ ∣ , fst ∥ ϕ ∥
```

Comment:
If `mon` was a record, then `mon-to-hom` could be a function defined in that record and thus simpler. In mathematical writing, mathematicians use "named fields", even if their formalism fakes it with tuples. The guide to formalization of set-theoretic material lies often in the informal text rather than in the set-encoding. The Agda then often comes out closer to practice!

Reply:
Sounds reasonable.  I'll make these changes now.